receiver/dcgmreceiver/util.go

```diff
@@ -19,39 +19,99 @@ package dcgmreceiver
-type rateIntegrator[V int64 | float64] struct {
-	lastTimestamp    int64
-	aggregatedRateUs V // the integration of the rate over microsecond timestamps.
+// For each metric, we need to track:
+type metricStats struct {
+	// Timestamp (us)
+	// Last value (for gauge metrics), as int or float
+	lastFieldValue *dcgm.FieldValue_v2
+	// Integrated rate (always int), as {unit-seconds,unit-microseconds}
+	integratedRateSeconds      int64
+	integratedRateMicroseconds int64
+	// Cumulative value (always int)
+	initialCumulativeValue int64
+	cumulativeValue        int64
+}
+
+func asInt64(fieldValue dcgm.FieldValue_v2) (int64, bool) {
+	// TODO: dcgm's Float64 and Int64 use undefined behavior
+	switch fieldValue.FieldType {
+	case dcgm.DCGM_FT_DOUBLE:
+		return int64(fieldValue.Float64()), true
+	case dcgm.DCGM_FT_INT64:
+		return fieldValue.Int64(), true
+	}
+	return 0, false
 }
 
-func (ri *rateIntegrator[V]) Reset() {
-	ri.lastTimestamp = nowUnixMicro()
-	ri.aggregatedRateUs = V(0)
+func asFloat64(fieldValue dcgm.FieldValue_v2) (float64, bool) {
+	switch fieldValue.FieldType {
+	case dcgm.DCGM_FT_DOUBLE:
+		return fieldValue.Float64(), true
+	case dcgm.DCGM_FT_INT64:
+		return float64(fieldValue.Int64()), true
+	}
+	return 0, false
 }
 
-func (ri *rateIntegrator[V]) Update(ts int64, v V) {
-	// Drop stale points.
-	if ts <= ri.lastTimestamp {
+func (m *metricStats) Update(fieldValue dcgm.FieldValue_v2) {
+	ts := fieldValue.Ts
+	intValue, intOk := asInt64(fieldValue)
+	if !intOk {
 		return
 	}
-	// v is the rate per second, and timestamps are in microseconds, so the
-	// delta will be 1e6 times the actual increment.
-	ri.aggregatedRateUs += v * V(ts-ri.lastTimestamp)
-	ri.lastTimestamp = ts
+	if m.lastFieldValue == nil {
+		m.initialCumulativeValue = intValue
+	} else {
+		if m.lastFieldValue.Ts >= ts {
+			return
+		}
+		m.cumulativeValue = intValue - m.initialCumulativeValue
+
+		tsDelta := ts - m.lastFieldValue.Ts
+		if fieldValue.FieldType == dcgm.DCGM_FT_DOUBLE {
+			m.integratedRateMicroseconds += int64(float64(tsDelta) * fieldValue.Float64())
+		} else {
+			m.integratedRateMicroseconds += tsDelta * intValue
+		}
+		m.integratedRateSeconds += m.integratedRateMicroseconds / 1000000
+		m.integratedRateMicroseconds %= 1000000
+	}
+	m.lastFieldValue = &fieldValue
 }
 
-func (ri *rateIntegrator[V]) Value() (int64, V) {
-	return ri.lastTimestamp, ri.aggregatedRateUs / V(1e6)
+type MetricsMap map[string]*metricStats
+
+func (m MetricsMap) LastFloat64(name string) (float64, bool) {
+	metric, ok := m[name]
+	if ok && metric.lastFieldValue != nil {
+		return asFloat64(*metric.lastFieldValue)
+	}
+	return 0, false
+}
+func (m MetricsMap) LastInt64(name string) (int64, bool) {
+	metric, ok := m[name]
+	if ok && metric.lastFieldValue != nil {
+		return asInt64(*metric.lastFieldValue)
+	}
+	return 0, false
+}
+func (m MetricsMap) IntegratedRate(name string) (int64, bool) {
+	metric, ok := m[name]
+	if ok {
+		return metric.integratedRateSeconds, true
```

Comment:
Just realized — this is fine for the ops agent, as its collection interval is always in whole seconds, but if we ever submit this upstream, nothing restricts the OTel collector interval, so this could be wrong. Let's add a comment documenting this assumption?

Reply:
Hm, I don't think this is wrong? It quantizes to the nearest unit-second, but since the original unit is per-second, that means it will just quantize to the nearest unit (e.g. bytes/second-seconds -> nearest byte). That seems correct to me regardless of whether the collection interval is in whole seconds, no? At worst it's just going to result in aliasing, but that's unavoidable if the cumulative is an INT64.

Reply:
Right, yes. It was confusing, so maybe we need a comment explaining this in more detail? Up to you.

Reply:
Comment added.